File: Data Structures & Algorithms/max-area-of-island/submission-3.cs
public class Solution {
    public int n, m;

    public int dfs(int r, int c, int[][] grid, bool[][] vis) {
        if (r < 0 || r >= n || c < 0 || c >= m || vis[r][c] || grid[r][c] == 0) {
            return 0;
        }

        vis[r][c] = true;

        int area = 1;

        area += dfs(r, c + 1, grid, vis);
        area += dfs(r, c - 1, grid, vis);
        area += dfs(r + 1, c, grid, vis);
        area += dfs(r - 1, c, grid, vis);

        return area;
    }

    public int MaxAreaOfIsland(int[][] grid) {
        n = grid.Length;
        m = grid[0].Length;

        bool[][] vis = new bool[n][];
        for (int i = 0; i < n; i++) {
            vis[i] = new bool[m];
        }

        int ans = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!vis[i][j] && grid[i][j] == 1) {
                    int area = dfs(i, j, grid, vis);
                    ans = Math.Max(ans, area);
                }
            }
        }

        return ans;
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
public:
    int dx[4] = {1, 0, -1, 0};
    int dy[4] = {0, 1, 0, -1};

    void dfs(int r, int c, vector<vector<char>>& grid, int n, int m) {
        if (r < 0 || c < 0 || r >= n || c >= m || grid[r][c] == '0')
            return;

        grid[r][c] = '0';

        for (int i = 0; i < 4; i++) {
            dfs(r + dx[i], c + dy[i], grid, n, m);
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty()) return 0;

        int n = grid.size();
        int m = grid[0].size();
        int ans = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {
                    dfs(i, j, grid, n, m);
                    ans++;
                }
            }
        }
        return ans;
    }
};